File: Java/93.java
__________________________________________________________________________________________________
sample 0 ms submission
class Solution {
    public List<String> restoreIpAddresses(String s) {
        List<String> ans = new ArrayList();
        if (s.length() > 3 && s.length() < 13)
            help(s, new char[s.length()+3], 0, ans, 0, 0);
        return ans;
    }
    
    void help(String s, char[] cs, int count, List<String> ans, int start, int i){
        int digit = 0;
        for (int l = start; l < 3+start && l < s.length(); l++){
            digit = 10*digit + (s.charAt(l)-'0');
            if (digit > 255) return;
            cs[i++] = s.charAt(l);
            if (count == 3){
                if (l+1==s.length()) ans.add(new String(cs));
            }else{
                cs[i] = '.';
                help(s, cs, count+1, ans, l+1, i+1);
            }
            if (digit == 0) return;
        }
    }
}
__________________________________________________________________________________________________
sample 34636 kb submission
public class Solution{
public List<String> restoreIpAddresses(String s) {
    List<String> result = new ArrayList<>();
    int[] path =new int[4];
    helper(result,s,0,path,0);
    return result;
}

void helper(List<String> acc, String s, int idx, int[] path,  int segment){  
    if(segment==4&idx==s.length()){
        acc.add(path[0]+"."+path[1]+"."+path[2]+"."+path[3]);
        return;
    }
    if(segment==4 || idx==s.length()){
        return;
    }
    for(int len=1;len<=3&&idx+len<=s.length();len++){
        int val=Integer.parseInt(s.substring(idx,idx+len));
        if(val>255||len >= 2  && s.charAt(idx) == '0'){
            break;
        }
        path[segment]=val;
        helper(acc,s,idx+len,path,segment+1);
        path[segment]=-1;
    }
 }
}
__________________________________________________________________________________________________

File: Python3/93.py
__________________________________________________________________________________________________
sample 28 ms submission
class Solution:
    def restoreIpAddresses(self, s: str) -> List[str]:
        res = set()
        def find(s, cb):
            if cb and len(cb)==4:
                if not s:
                    res.add('.'.join(cb))
                return
            if not s:
                return
            find(s[1:], cb+[s[:1]])
            if s[0]=='0':
                return
            n = len(s)
            if n>=2:
                find(s[2:], cb+[s[:2]])
            if n>=3:
                val = int(s[:3])
                if val<=255:
                    find(s[3:], cb+[s[:3]])
        find(s, [])
        return list(res)
__________________________________________________________________________________________________
sample 12968 kb submission
class Solution:
    # @param s: str
    # @return List[str]
    def restoreIpAddresses(self, s):
        results = []
        self.dfs(results, s, "", 0, 1)
        return results
    
    def dfs(self, results, s, path, start, section):
        if section == 4:
            segment = s[start:]
            if self.isValid(segment):
                results.append(path + segment)
            return None
        elif section > 4:
            return None
        for i in range(start, start + 3):
            segment = s[start:i + 1]
            if self.isValid(segment):
                self.dfs(results, s, path + segment + ".", i + 1, section + 1)
        return None
    
    # @param s: str
    # @return bool
    def isValid(self, s):
        if len(s) > 3 or len(s) == 0:
            return False
        
        if len(s) == 1 and s[0] == "0":
            return True
        elif s[0] == "0":
            return False
        
        return 0 < int(s) <= 255
__________________________________________________________________________________________________
sample 13072 kb submission
class Solution:
    def restoreIpAddresses(self, s: str) -> List[str]:
        def dfs(i, path):
            nonlocal ans
            if len(path) >= 4 and i < len(s): return
            if i >= len(s):
                if len(path) == 4:
                    ans.append('.'.join(path))
                return
            
            for j in range(i + 1, min(len(s) + 1, i + 4)):
                k = int(s[i: j])
                if s[i] == '0' and j > i + 1: continue
                if 0 <= k <= 255:
                    dfs(j, path + [str(k)])
                    
        ans = []
        dfs(0, [])
        return ans